rpc/personal_api.go

```diff
@@ -1,30 +1,168 @@
+	defer e.keybaseLock.Unlock()
+
+	if e.cliCtx.Keybase == nil {
+		keybase, err := keyring.NewKeyring(
+			sdk.KeyringServiceName(),
+			viper.GetString(flags.FlagKeyringBackend),
+			viper.GetString(flags.FlagHome),
+			e.cliCtx.Input,
+			keyring.WithKeygenFunc(emintcrypto.EthermintKeygenFunc),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		e.cliCtx.Keybase = keybase
+	}
+
+	return e.cliCtx.Keybase.List()
+}
+
+// ImportRawKey stores the given hex encoded ECDSA key into the key directory,
+// encrypting it with the passphrase.
+func (e *PersonalEthAPI) ImportRawKey(privkey, password string) (common.Address, error) {
+	return common.Address{}, nil
+}
+
+// ListAccounts will return a list of addresses for accounts this node manages.
+func (e *PersonalEthAPI) ListAccounts() ([]common.Address, error) {
+	addrs := []common.Address{}
+	infos, err := e.getKeybaseInfo()
+	if err != nil {
+		return addrs, err
+	}
+
+	for _, info := range infos {
+		addressBytes := info.GetPubKey().Address().Bytes()
+		addrs = append(addrs, common.BytesToAddress(addressBytes))
+	}
+
+	return addrs, nil
+}
+
+// LockAccount will lock the account associated with the given address when it's unlocked.
+func (e *PersonalEthAPI) LockAccount(address common.Address) bool {
+	for i, key := range e.keys {
+		if bytes.Equal(key.PubKey().Address().Bytes(), address.Bytes()) {
+			tmp := make([]emintcrypto.PrivKeySecp256k1, len(e.keys)-1)
+			copy(tmp[:i], e.keys[:i])
+			copy(tmp[i:], e.keys[i+1:])
+			e.keys = tmp
+			return true
+		}
+	}
+
+	return false
+}
+
+// NewAccount will create a new account and returns the address for the new account.
+func (e *PersonalEthAPI) NewAccount(password string) (common.Address, error) {
+	_, err := e.getKeybaseInfo()
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// TODO: is keyring.Secp256k1 the correct algorithm?
+	info, _, err := e.cliCtx.Keybase.CreateMnemonic("key_"+time.Now().Format(time.RFC3339), keyring.English, password, keyring.Secp256k1)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return common.BytesToAddress(info.GetPubKey().Address().Bytes()), nil
+}
+
+// UnlockAccount will unlock the account associated with the given address with
+// the given password for duration seconds. If duration is nil it will use a
+// default of 300 seconds. It returns an indication if the account was unlocked.
+func (e *PersonalEthAPI) UnlockAccount(ctx context.Context, addr common.Address, password string, duration *uint64) (bool, error) {
+	// TODO: use duration
+
+	infos, err := e.getKeybaseInfo()
+	if err != nil {
+		return false, err
+	}
+
+	name := ""
+	for _, info := range infos {
+		addressBytes := info.GetPubKey().Address().Bytes()
+		if bytes.Equal(addressBytes, addr[:]) {
+			name = info.GetName()
+		}
+	}
+
+	if name == "" {
+		return false, fmt.Errorf("cannot find key with given address")
+	}
+
+	// TODO: this only works on local keys
+	privKey, err := e.cliCtx.Keybase.ExportPrivateKeyObject(name, password)
```

Comment:
I don't understand why we are exporting the privkey in order to unlock it

Reply:
it's a bit of a workaround but the "unlocked" keys are stored in `e.keys`, so to unlock the key we need to export it and save it in `e.keys`, let me know if this isn't a good idea

Reply:
hmm I guess it's ok for now? We should document it on the comments and create an issue to track this work